index.js

```diff
@@ -366,31 +371,61 @@ AFRAME.registerComponent('teleport-controls', {
     var intersects = this.raycaster.intersectObjects(meshes, true);
     if (intersects.length > 0 && !this.hit &&
         this.isValidNormalsAngle(intersects[0].face.normal)) {
+      // save the hit point and then return true
       var point = intersects[0].point;
-
-      this.line.material.color.set(this.curveHitColor);
-      this.line.material.opacity = this.data.hitOpacity;
-      this.line.material.transparent= this.data.hitOpacity < 1;
-      this.hitEntity.setAttribute('position', point);
-      this.hitEntity.setAttribute('visible', true);
-
-      this.hit = true;
       this.hitPoint.copy(intersects[0].point);
+      return true;
+    }
+    return false;
+  },
 
-      // If hit, just fill the rest of the points with the hit point and break the loop
-      for (var j = i; j < this.line.numPoints; j++) {
-        this.line.setPoint(j, this.hitPoint);
+  /**
+   * Set the line materials for rendering
+   * @param {boolean} isCollided to determine the hit/miss material of the line
+  */
+  setLineMaterial: function (isCollided) {
+    const color = isCollided ? this.curveHitColor : this.curveMissColor;
+    const opacity = isCollided ? this.data.hitOpacity : this.data.missOpacity;
+    const transparent = isCollided ? (this.data.hitOpacity < 1) : (this.data.missOpacity < 1);
+    this.line.material.color.set(color);
+    this.line.material.opacity = opacity;
+    this.line.material.transparent = transparent;
+  },
+
+  /**
+   * Set line points for rendering
+   *
+  */
+  setLinePoints: function () {
+    const numPoints = this.raycastPoints.length;
+    for (let i = 0; i < numPoints; i++) {
+      if (this.data.drawIncrementally && i >= this.numActivePoints) {
+        this.line.setPoint(i, this.raycastPoints[this.numActivePoints]);
+      } else {
+        this.line.setPoint(i, this.raycastPoints[i]);
       }
-      return true;
-    } else {
-      this.line.setPoint(i, next);
-      return false;
     }
   },
 
-  isValidNormalsAngle: function (collisionNormal) {
-    var angleNormals = this.referenceNormal.angleTo(collisionNormal);
-    return (THREE.Math.RAD2DEG * angleNormals <= this.data.landingMaxAngle);
+  /**
+   * update hitEntity & line material while there's a collision.
+   * only called while there's a collision
+  */
+  updateLineAndHitEntityByCollisions: function () {
```

Comment:
Right now the hit entity looks bright green even before the line reaches it. Let's change that, since the user isn't supposed to be able to teleport until the line reaches the hitEntity. Perhaps the hitEntity should be especially transparent until teleportation is allowed? Or it can be a different color? Perhaps some effect can be animated on the hitEntity while the line is shooting out towards it?